### PR TITLE
App refactoring

### LINKED ISSUE #398: Application refactoring
The initial and current design of the ct-app is, among other things, based on a latency measure between ct-nodes and peers in the wild.
Now that latency measure has been completely bypassed, the role of the Aggregator, AggregatorTrigger, metricTable and some part of the Netwatcher is not that clear anymore.

A great simplification of the app could be done by giving more responsibilities to the actual EconomicHandler, still relying on multiple ct nodes to handle the messages load.

The new form of ct-app would be composed only of: 1 EconomicHandler, X Postman,  1 FeedbackPostman, 1 rewardTable.

This simplification would be helpful when coming to migrating the app to Rust.

### Related issues
The following issues will be addressed when refactoring the app:
- [x] #207 
- [ ] #211
- [x] #266 
- [x] #311 
- [x] #349
- [x] #363 
- [x] #378 
- [x] #397 
- [ ] #404
- [ ] (#406)

### PR DESCRIPTION
**Pull request by jeandemeusy**
This PR refactors completely the ct-dapp.

**TL;DR**: Before, the app was composed of many modules communicating together. In this PR, many of them are integrated within the same module. Precisely, the previous `Netwatcher`, `Aggregator`, `AggregatorTrigger` and `EconomicHandler` (and a database table `metricTable`) are now united in the module called `Core`.

### Precisions
- #207
When running the app, two environment variables can be specified regarding subgraph URL:
  - `PARAM_SUBGRAPH_URL` is mandatory
  - `PARAM_SUBGRAPH_URL_BACKUP` is optional

Regardless of the second one being specified, the app will try from time to time to make a request to the available URLs. The URL used by the app is set using this preferred order: `default`, `backup`, `None`.

- #266

In the refactored version of the app, every variable that needs to be thread safe is a `LockedVar` instance. Testing locks is then super simple, as testing this class will ensure that locking works as expected.

- #311
 
This balance calculation refers to the total balance in channels for every given peers. This was done in the `HoprdAPIHelper` class. This calculation has now been simplified and moved out of the API helper to the `Utils` class.

- #363

Each time the economic model is applied (not reward distributed), a `NEXT_DISTRIBUTION_S`  Prometheus metric is updated to store the next distribution epoch. Also, `DISTRIBUTION_DELAY` Prometheus metric now exposes the delay between two distribution.

- #378

Every mention to APY is modified to APR.

- #397

This has been fixed by using a unique tag for each and every peer. To be sure to have a unique tag, a peerTable has been created. When sending a message to this peer, the peer is either added to or retrieved from this table, and the tag used is 800 + peer.id.

- #404

Total funding that went into the node's safes is now measured by summing results from a subgraph request. The total funding is exposed on Prometheus (Grafana). From this value, if we subtract the total stake that our nodes holds (in safe + in channels), also exposed on Prometheus (Grafana), we get the total value distributed through CT.
If needed, the final calculation (`total_funding - sum(node_stake)`) can be done directly on Grafana.

- #406

This issue not has been tackled as it seams too costly to do. If at every distribution, to every single peer we fund (on-chain operation) all the channels, this would mean 400 peers * 4 nodes * 4 distribution = 6'400 on-chain operation/day. Realistically, it's way cheaper to have channels opened with a higher balance, and to top up when needed. However, the threshold value that triggers funding of a channel needs to be higher than the maximum value someone is expecting to receive for a given distribution, which is the case (funding triggers when balance goes bellow 15HOPR, no one is eligible for this much at a single distribution). This ensures that every peer will be able to redeem ticket from, at least, the last distribution.

Resolves #398

### CODE REVIEW

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -38,10 +38,10 @@ jobs:
...
           ruff --output-format=github --ignore E501 --target-version=py39 .
 
-      - name: Test with pytest
-        working-directory: ct-app
-        run: |
-          pytest tests -k "not api_helper.py and not gcp"
+      # - name: Test with pytest
+      #   working-directory: ct-app
+      #   run: |
+      #     pytest tests -k "not api_helper.py and not gcp"
```
Comment (Teebor-Choka) [inline .github/workflows/main.yaml:44]: Careful about uncommenting this before merge.

File: ct-app-unified/core/model/subgraph_type.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    @classmethod
+    def callables(cls):
+        return [item for item in cls if item != cls.NONE]
+
+    def toInt(self):
+        if self == SubgraphType.DEFAULT:
+            return 0
+        if self == SubgraphType.BACKUP:
+            return 1
+        return -1
```
Comment (Teebor-Choka) [inline ct-app-unified/core/model/subgraph_type.py:18]:
```suggestion
        return { SubgraphType.DEFAULT: 0, SubgraphType.BACKUP: 1 }.get(self, -1)
```

File: ct-app-unified/database/database_connection.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    This class requires the following environment variables to be set:
+    - PGUSER: the database user name
+    - PGPASSWORD: the database user password
+    - PGHOST: the database host
+    - PGPORT: the database port
+    - PGDATABASE: the database name
+    - PGSSLMODE: the SSL mode
+    - PGSSLROOTCERT: the SSL root certificate
+    - PGSSLCERT: the SSL certificate
+    - PGSSLKEY: the SSL key
```
Comment (Teebor-Choka) [inline ct-app-unified/database/database_connection.py:26]: It's better to assert this inside the init function than have a comment. Also easier to debug.

File: ct-app-unified/test/components/test_decorators.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+@pytest.fixture
+def foo_class():
+    return FooClass()
+
+
+@pytest.mark.asyncio
+async def test_connectguard(foo_class: FooClass):
+    await foo_class.foo_connectguard_func()
+    pytest.skip("Not implemented")
```
Comment (Teebor-Choka) [inline ct-app-unified/test/components/test_decorators.py:39]: Is it going to be tested?

File: .github/workflows/main.yaml [hunk 5]
```diff
@@ -27,21 +27,21 @@ jobs:
           python-version: 3.9
 
       - name: Install dependencies
-        working-directory: ct-app
+        working-directory: ct-app-unified
```
[the remark this answers is not in the document]
Comment (jeandemeusy) [reply .github/workflows/main.yaml:30]: I started in a different directory as the changes were quite big, and I wanted to see this as two distinct app. But completely makes sense to replace the old implementation by the new one, and keep the same working dir in the workflow.

File: ct-app-unified/core/__main__.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    start_http_server(8080)
+
+    loop = asyncio.new_event_loop()
+    loop.add_signal_handler(SIGINT, instance.stop)
+    loop.add_signal_handler(SIGTERM, instance.stop)
+
+    try:
+        loop.run_until_complete(instance.start())
+    except asyncio.CancelledError:
+        instance._error("Stopping the instance...")
```
[the remark this answers is not in the document]
Comment (jeandemeusy) [reply ct-app-unified/core/__main__.py:36]: Private method that "should" not be used outside the class. However, this error statement I didn't have the choice to access it from outside.

File: ct-app-unified/test/components/test_decorators.py (hunk 4 above)
Comment (jeandemeusy) [reply ct-app-unified/test/components/test_decorators.py:39]: Yep ! I added a bunch of method in `utils` that should be tested as well with some decorators.

File: ct-app-unified/database/database_connection.py (hunk 3 above)
Comment (jeandemeusy) [reply ct-app-unified/database/database_connection.py]: Done in commit [5d257a2](https://github.com/hoprnet/ct-research/pull/409/commits/5d257a289cc2393a66dfcbfd5a7a7afd63325480)

File: ct-app-unified/core/__main__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeandemeusy) [reply ct-app-unified/core/__main__.py]: Underscore removed in all print related statement in [60e1f16](https://github.com/hoprnet/ct-research/pull/409/commits/60e1f169167e99b93c5272bacfba27ac4bb5f517)

File: .github/workflows/main.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeandemeusy) [reply .github/workflows/main.yaml:30]: 'ct-app' removed in [d8a5a78](https://github.com/hoprnet/ct-research/pull/409/commits/d8a5a78e265d924a3cb4ccee66bd70d6b1d3b2ad) and 'ct-app-unified' renamed to 'ct-app'

File: ct-app-unified/test/components/test_decorators.py (hunk 4 above)
Comment (jeandemeusy) [reply ct-app-unified/test/components/test_decorators.py]: Decorator tested in [4973614](https://github.com/hoprnet/ct-research/pull/409/commits/49736149533c125ba8c48c18e065ecd0007b9aa7) (small fix in [a6ba661](https://github.com/hoprnet/ct-research/pull/409/commits/a6ba6616189ee9fcb5658023b1b1ffb88e9b514a)). Missing test on `Utils` methods will be added in a separate PR for simplicity. Code already running in staging using those methods

**Review by Teebor-Choka — APPROVED** [review]
